### ISSUE TITLE
systemd-networkd-wait-online gets stuck with multiple interfaces

### ISSUE THREAD

**Issue by danijelt** [systemd/systemd#5154]
### Submission type

  - [x] Bug report

### systemd version the issue has been seen with

231 release version (25 July 2016)

### Used distribution

custom, made with Buildroot

### In case of bug report: Expected behaviour you didn't see

According to the documentation, systemd-networkd-wait-online "will wait for all links it is aware of and which are managed by systemd-networkd.service(8) to be fully configured or failed, and for at least one link to gain a carrier".  

### In case of bug report: Unexpected behaviour you saw

However, if two network interfaces are present (eth0 and wlan0) and one of them can't gain carrier (either eth0 unplugged or wlan0 can't find the configured network) and both are in "up" state (checked with `ip a`), systemd-networkd-wait-online will keep waiting for the second interface to gain carrier, contrary to the documentation.  
Since I must have network on boot, I have `--timeout=0` which results in stuck boot process.

### In case of bug report: Steps to reproduce the problem

* Configure network interfaces in `/etc/systemd/network/`, either with `Match=*` or separately for eth0 and wlan0. **Both behave the same regarding this bug**.

* Create `/etc/systemd/system/systemd-networkd-wait-online.service`. Add `--timeout=0` in ExecStart. `--interface=eth0 --interface=wlan0` **can be set, but doesn't have to be, it doesn't change anything regarding this bug**.

* Create an unit that requires `systemd-networkd-wait-online` (if it doesn't exist) and reboot with either eth0 unplugged or with wlan0 set to connect to an out-of-range (or invalid) network. `wpa_supplicant` is used for configuring wlan0 as explained in documentation.

* `systemd-networkd-wait-online` will now wait forever until both interfaces gain carrier, and unit that requires it won't be started.

### Additional - systemd units and configuration
My configuration and units:
```
# cat /etc/systemd/network/wired.network 
[Match]
Name=eth0

[Network]
DHCP=ipv4

[DHCP]
ClientIdentifier=mac

# cat /etc/systemd/network/wireless.network 
[Match]
Name=wlan0

[Network]
DHCP=ipv4

[DHCP]
ClientIdentifier=mac

# cat /etc/systemd/system/systemd-networkd-wait-online.service 
[Unit]
DefaultDependencies=no
Conflicts=shutdown.target
Requisite=systemd-networkd.service
After=systemd-networkd.service
Before=network-online.target
JobTimeoutSec=infinity

[Service]
Type=oneshot
TimeoutSec=infinity
ExecStart=/lib/systemd/systemd-networkd-wait-online --interface=eth0 --interface=wlan0 --timeout=0
RemainAfterExit=yes

# cat /etc/systemd/system/wpa_supplicant.service 
[Unit]
Description=WPA supplicant daemon
Wants=boot.mount
JobTimeoutSec=10

[Service]
Type=simple
ExecStart=/sbin/wpa_supplicant -c/boot/wpa_supplicant.conf -iwlan0

# cat /boot/wpa_supplicant.conf 
network={
	ssid="my ssid"
	psk=key generated with wpa_passphrase
}
```

**Comment by gdamjan** [comment]
possibly related, in latest master (74d8f1c55):
```
$ sudo networkctl 
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           routable    configured
  2 ens5             ether              routable    configured
  3 bridge0          ether              no-carrier  configuring
```
```
$ sudo /usr/lib/systemd/systemd-networkd-wait-online 
ignoring: lo
Event loop failed: Connection timed out
```
brdige0 is in state **configuring** because it has a .network file with just a [Match] section (to bring it up).

ps.
this works though:
```
$ sudo /usr/lib/systemd/systemd-networkd-wait-online -i ens5
ignoring: bridge0
ignoring: lo
```

**Comment by de-vri-es** [comment]
It would be nice to be able to wait for any interface in a set of interfaces. I have a machine with two ethernet interfaces, both of which are fine to use. One is configured to use DHCP and the other has a static configuration. It would be nice if there was something like:
```
systemd-networkd-wait-online --any -i interface0 -i interface1
systemd-networkd-wait-online --any interface0 interface1          # or this syntax
```

Expected behaviour would be for `systemd-networkd-wait-online` to return success when *any* of the listed interfaces is up and configured.

**Comment by martyg77** [comment]
Appears related to what I am seeing in #2154

**Comment by sitsofe** [comment]
Might be related to #3752 (systemd-networkd can't configure interface if dhcp server does not provide routers)

**Comment by bobrik** [comment]
Seeing this too.

```
# networkctl | cat
IDX LINK             TYPE               OPERATIONAL SETUP
  1 lo               loopback           carrier     unmanaged
  2 eth0             ether              off         unmanaged
  3 eth1             ether              off         unmanaged
  4 eth2             ether              carrier     configuring
  5 eth3             ether              no-carrier  configuring
  6 bond0            ether              degraded    configured
  7 vlan99           ether              routable    configured
  8 vlan999          ether              routable    configured
```

It works if I exclude `eth2` and `eth3` (slaves of `bond0`):

```
# /lib/systemd/systemd-networkd-wait-online --timeout 3 --ignore eth2 --ignore eth3; echo $?
ignoring: eth3
ignoring: eth2
ignoring: lo
0
```

Both vlan interfaces are on top of `bond0`.

**Comment by graysky2** [comment]
I'm seeing this on a dual network setup as well (wlan0 and eth0).  I have a service that requires a network connection and I want to make it wait for one before starting.... is the recommended way to handle this to simply copy `/usr/lib/systemd/system/systemd-networkd-wait-online.service` to `/etc/systemd/system/systemd-networkd-wait-online.service` and to modify ExecStart= to read something like `ExecStart=/usr/lib/systemd/systemd-networkd-wait-online --ignore eth0`

??

**Comment by Scrumplex** [comment]
I am seeing this behavior on a server with a few virtual interfaces. 
```
root@central-nbg1-1 ~# networkctl
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           carrier     unmanaged 
  2 ens3             ether              routable    configuring
  3 tun0             none               routable    unmanaged 
  4 docker0          bridge             no-carrier  unmanaged 

4 links listed.
```

Some of my services do not start as they are waiting for the network, although it is reachable while systemd-networkd-wait-online is waiting.

**Comment by Scrumplex** [comment]
This may be related to #1850. My main interface (ens3) fails on boot:
```
Jan 22 17:08:47 central-nbg1-1 systemd-networkd[541]: ens3: Gained carrier
Jan 22 17:08:47 central-nbg1-1 systemd-networkd[541]: ens3: Could not set route: Network is unreachable
```
This may trick networkd into believing that the iface is not up, although the iface works later in the boot process.

**Comment by jojoax** [comment]
Fails up to now, 2019-04-24 
[root@pc3 ~]# /usr/lib/systemd/systemd-networkd-wait-online --timeout=10 --any -i br0 
Event loop failed: Connection timed out

Even br0 is operational and correctly working.

**Comment by jojoax** [comment]
Workaround for buggy networkd-systemd-wait-online service:
1. install fping
2. write a script to ping your gateway, nfs server, dhcp server like this:
#!/bin/bash
while true; do /usr/bin/fping -t 150 -i 10 192.168.11.10 -q && break; done

3. copy  systemd-networkd-wait-online.service into /etc/systemd/system
change the  ExecStart= line to use your script. 

4. Afterwards all dependencies from systemd-networkd-wait-online.service are working find until this ugly bug is fixed

**Comment by yuwata** [comment]
@jojoax Please test with v242 if you do not use v242. And please provide the follwoing:
- .network and .netdev files
- debugging logs of `systemd-networkd`. The debugging log can be generated by creating the following file.
```
# /etc/systemd/system/systemd-networkd.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```
- Also, please try to run `wait-online` with debug option:
```
# SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/systemd-networkd-wait-online
```

**Comment by jojoax** [comment]
Here you are:

[01-krannetz-bridge.netdev.txt](https://github.com/systemd/systemd/files/3115425/01-krannetz-bridge.netdev.txt)
[02-krannetz.network.txt](https://github.com/systemd/systemd/files/3115426/02-krannetz.network.txt)
[03-krannetz.network.txt](https://github.com/systemd/systemd/files/3115427/03-krannetz.network.txt)
[systemd-networkd-wait-online.service.txt](https://github.com/systemd/systemd/files/3115431/systemd-networkd-wait-online.service.txt)

`
[root@pc3 system]# systemctl start  systemd-networkd-wait-online.service 
Job for systemd-networkd-wait-online.service failed because the control process exited with error code.
See "systemctl status systemd-networkd-wait-online.service" and "journalctl -xe" for details.
[root@pc3 system]# systemctl status  systemd-networkd-wait-online.service 
● systemd-networkd-wait-online.service - Wait for Network to be Configured
   Loaded: loaded (/etc/systemd/system/systemd-networkd-wait-online.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2019-04-25 08:17:09 CEST; 7s ago
     Docs: man:systemd-networkd-wait-online.service(8)
  Process: 3510 ExecStart=/usr/lib/systemd/systemd-networkd-wait-online --timeout=15 --ignore lo --ignore eth0 --ignore eth1 --ignore eth2 -i kran0 (code=exited, status=1/>
  Process: 3511 ExecStopPost=/usr/bin/networkctl status -a (code=exited, status=0/SUCCESS)
 Main PID: 3510 (code=exited, status=1/FAILURE)

Apr 25 08:17:09 pc3 networkctl[3511]:     Network File: /etc/systemd/network/02-krannetz.network
Apr 25 08:17:09 pc3 networkctl[3511]:             Type: ether
Apr 25 08:17:09 pc3 networkctl[3511]:            State: no-carrier (configuring)
Apr 25 08:17:09 pc3 networkctl[3511]:             Path: pci-0000:02:00.1
Apr 25 08:17:09 pc3 networkctl[3511]:           Driver: igb
Apr 25 08:17:09 pc3 networkctl[3511]:           Vendor: Intel Corporation
Apr 25 08:17:09 pc3 networkctl[3511]:            Model: 82576 Gigabit Network Connection (Gigabit ET Dual Port Server Adapter)
Apr 25 08:17:09 pc3 networkctl[3511]:       HW Address: 68:05:ca:8f:2b:67 (Intel Corporate)
Apr 25 08:17:09 pc3 systemd[1]: systemd-networkd-wait-online.service: Failed with result 'exit-code'.
Apr 25 08:17:09 pc3 systemd[1]: Failed to start Wait for Network to be Configured.
`

PS: the webserver is shit, refusing attaching the files in original

**Comment by yuwata** [comment]
> State: no-carrier (configuring)

The result of `networkctl` is truncated, but it seems all the bridge slaves are in no-carrier state. no?

**Comment by jojoax** [comment]
No, one of three is link up and working.

**Comment by yuwata** [comment]
Hmm. Please provide debugging logs.

**Comment by jojoax** [comment]
[root@pc3 system]# ethtool eth0
Settings for eth0:
        Supported ports: [ TP ]
...
        Link detected: yes
[root@pc3 system]# ethtool eth1
Settings for eth1:
        Supported ports: [ TP ]
...
        Link detected: yes
[root@pc3 system]# ethtool eth2
Settings for eth2:
        Supported ports: [ TP ]
...
        Link detected: no
SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/systemd-networkd-wait-online --timeout=15 --ignore lo --ignore eth0 --ignore eth1 --ignore eth2 -i kran0 
Found link 5
Found link 4
Found link 3
Found link 2
Found link 1
kran0: Operational state 'degraded-carrier' is below 'degraded'
Event loop failed: Connection timed out
[root@pc3 system]# echo $?
1


Degraded is somewhat ok, but this bridge is intended use as "switch", so one (or more) failing ports may be ok, (execpt missing uplink, but this is not scope here)

**Comment by yuwata** [comment]
I see, then please try `-i kran0:degraded-carrier`.

If it does not work, then please provide debugging logs of `systemd-networkd` (not `systemd-networkd-wait-online`). The debugging log can be generated by creating the following file.
```
# /etc/systemd/system/systemd-networkd.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```

**Comment by yuwata** [comment]
Let's close this. If you still have problems, then please file as a new issue with debugging logs.
